### PR TITLE
move Pkg to stdlib

### PR DESCRIPTION
**Pull request by KristofferC**
Pkg3 will replace Pkg(2) but moving it to the stdlib first will make that transition easier.

On the few places Pkg is used in base, I temporarily just used a Ref that Pkg fills.

### CODE REVIEW

File: base/interactiveutil.jl [hunk 1]
```diff
@@ -340,11 +340,11 @@ function versioninfo(io::IO=STDOUT; verbose::Bool=false, packages::Bool=false)
     end
     if packages || verbose
         println(io, "Packages:")
-        println(io, "  Package Directory: ", Pkg.dir())
+        println(io, "  Package Directory: ", PKG_MODULE_REF[].dir())
         print(io, "  Package Status:")
-        if isdir(Pkg.dir())
+        if isdir(PKG_MODULE_REF[].dir())
```
Comment (vchuravy) [inline base/interactiveutil.jl]: I think you might need to use `invokelatest` here. @vtjnash should be able to confirm that.

File: stdlib/Pkg/src/Pkg.jl [hunk 2]
```diff
@@ -13,9 +13,9 @@ Please see the manual section on packages for more information.
 """
 module Pkg
 
-export Dir, Types, Reqs, Cache, Read, Query, Resolve, Write, Entry
-export dir, init, rm, add, available, installed, status, clone, checkout,
-       update, resolve, test, build, free, pin, PkgError, setprotocol!
+#export Dir, Types, Reqs, Cache, Read, Query, Resolve, Write, Entry
```
Comment (vchuravy) [inline stdlib/Pkg/src/Pkg.jl]: These should probably be kept exported?

File: base/interactiveutil.jl (hunk 1 above)
Comment (KristofferC) [reply base/interactiveutil.jl]: Seems to work ok actually.

**Comment by KristofferC** [comment]
So this made the Pkg tests start writing their info messages to STDOUT for some reason...

Also, the error message extremely unhelpful:

```
Pkg: Error During Test at nothing:1
  Test threw an exception of type RemoteException
  Expression: Pkg
  Error deserializing a remote exception from worker 2
  Remote(original) exception of type TestSetException
  Remote stacktrace : 
  finish at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Test/src/Test.jl:784
  macro expansion at ./util.jl:310 [inlined]
  top-level scope at ./<missing>:19
  eval at ./boot.jl:295
  #runtests#1 at /usr/home/julia/worker/11rel-amd64/build/test/testdefs.jl:25
  #runtests at ./<missing>:0 [inlined] (repeats 2 times)
  #125 at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:269
  run_work_thunk at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:56
  macro expansion at /usr/home/julia/worker/11rel-amd64/build/usr/share/julia/site/v0.7/Distributed/src/process_messages.jl:269 [inlined]
  #124 at ./event.jl:92
```

**Comment by KristofferC** [comment]
I can't reproduce the 1 test error locally and the error message says absolutely nothing.

**Comment by JeffBezanson** [comment]
Looks like freebsd is still trying to run `pkg` tests?

**Comment by KristofferC** [comment]
Yeah, perhaps that can be removed @iblis?

Does anyone know the reason for splitting the tests into two parts for CI? Something with online / offline tests but `choosetests` already checks Internet connectivity and filters based on that.

File: .travis.yml [hunk 3]
```diff
@@ -144,7 +144,7 @@ script:
     - /tmp/julia/bin/julia -e 'versioninfo()'
     - pushd /tmp/julia/share/julia/test
     - /tmp/julia/bin/julia --check-bounds=yes runtests.jl $TESTSTORUN &&
-      /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download pkg
+      /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download
```
Comment (iblis17) [inline .travis.yml]: `.circleci/config.yml` need to be changed as well?

**Comment by iblis17** [comment]
I think I should provde cumstom shell script like Travis and CircleCI does in long-term...
(Let me figure out how to make it

File: .travis.yml (hunk 3 above)
Comment (KristofferC) [reply .travis.yml]: Yes, I didn't know that existed, thanks.

**Comment by KristofferC** [comment]
All CI now fail with 

```
julia: /home/travis/build/JuliaLang/julia/src/codegen.cpp:1096:
jl_llvm_functions_t jl_compile_linfo(jl_method_instance_t**, jl_code_info_t*, size_t, const jl_cgparams_t*):
Assertion `li->min_world <= world && li->max_world >= world' failed.
```

During precompile. Perhaps adding some debug printing to `precompile.jl` would be useful?

**Comment by StefanKarpinski** [comment]
cc @vtjnash ^

**Comment by vchuravy** [comment]
I would suspect that the worldage error is: https://github.com/JuliaLang/julia/pull/25705#discussion_r163283315

File: .freebsdci.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    export MALLOC_CONF='junk:false'
+    export VERBOSE=1
+    export FORCE_ASSERTIONS=1
+    export LLVM_ASSERTIONS=1
+    export JULIA_TEST_MAXRSS_MB=600
+    export JULIA_CPU_CORES=$MAKE_JOBS_NUMBER
+
+    gmake testall \
+        test-download \
+        test-Pkg \ # <- TODO: Should only test Pkg/pkg but no rule for test-Pkg/pkg
```
Comment (iblis17) [inline .freebsdci.sh]: The comment in this line neutralized the effect of `\` , so FreeBSD CI build failed.

line 73:
https://freebsdci.julialang.org/#/builders/1/builds/7031/steps/7/logs/stdio

**Comment by KristofferC** [comment]
@vchuravy But the `PKG_MODULE_REF` thing doesn't exist at all anymore on the newer commits because enough stuff got moved to stdlib that it wasn't needed anymore.

File: .freebsdci.sh (hunk 4 above)
Comment (KristofferC) [reply .freebsdci.sh]: Oops, thanks.

Comment (KristofferC) [reply .freebsdci.sh]: Is there anyway we could run this as `julia test/runtests.jl test-Pkg/pkg`?

Comment (iblis17) [reply .freebsdci.sh]: I think invoking `./usr/bin/julia` should work

File: .freebsdci.sh [hunk 5]
```diff
@@ -36,10 +36,8 @@ runtests(){
     export JULIA_TEST_MAXRSS_MB=600
     export JULIA_CPU_CORES=$MAKE_JOBS_NUMBER
 
-    gmake testall \
-        test-download \
-        test-pkg \
-        test-libgit2-online
+    ./usr/bin/julia test all
```
Comment (KristofferC) [inline .freebsdci.sh]: @iblis17 Did this mess up for the builder?

File: .freebsdci.sh [hunk 6]
```diff
@@ -36,10 +36,8 @@ runtests(){
     export JULIA_TEST_MAXRSS_MB=600
     export JULIA_CPU_CORES=$MAKE_JOBS_NUMBER
 
-    gmake testall \
-        test-download \
-        test-pkg \
-        test-libgit2-online
+    ./usr/bin/julia test/runtests.jl all
+    ./usr/bin/julia test/runtests.jl libgit2-online Pkg/pkg download
```
Comment (iblis17) [inline .freebsdci.sh]: need `--check-bounds=yes`?

Comment (KristofferC) [reply .freebsdci.sh]: Don't think so, already gets applied at https://github.com/JuliaLang/julia/blob/master/test/testenv.jl#L20

Comment (iblis17) [reply .freebsdci.sh]: 👌

**Comment by KristofferC** [comment]
The world error assertion was disabled from consulting `@vtjnash` and he said it wasn't really a problem disabling it and that he would look at it later.

**Comment by StefanKarpinski** [comment]
giphy-1

**Pull request merged**